runtime/lua/vim/lsp/util.lua

```diff
@@ -1290,18 +1290,169 @@ function M._trim(contents, opts)
-    end
-  end
-  return fences
+---@alias MarkdownBlock {line:string}
+---@alias MarkdownCodeBlock {code:string[], lang:string}
+---@alias Markdown (MarkdownBlock|MarkdownCodeBlock)[]
+
+---@param text string
+---@param opts? {ft?: string}
+---@return Markdown
+local function parse_markdown(text, opts)
+  opts = opts or {}
+  text = text:gsub('</?pre>', '```'):gsub('\r', '')
+
+  -- Replace common HTML entities.
+  local entities =
+    { nbsp = '', lt = '<', gt = '>', amp = '&', quot = '"', apos = "'", ensp = ' ', emsp = ' ' }
+  for entity, char in pairs(entities) do
+    text = text:gsub('&' .. entity .. ';', char)
+  end
+
+  local ret = {} ---@type Markdown
+  local lines = vim.split(text, '\n')
+  local l = 1
+
+  local function is_empty(line)
+    return line and line:find('^%s*$')
+  end
+
+  local function eat_nl()
+    while is_empty(lines[l + 1]) do
+      l = l + 1
+    end
+  end
+
+  local function is_rule(line)
+    return line and line:find('^%s*[%*%-_][%*%-_][%*%-_]+%s*$')
+  end
+
+  local function is_code_block(line)
+    return line and line:find('^%s*```')
+  end
+
+  while l <= #lines do
+    local line = lines[l]
+    if is_empty(line) then
+      local is_start = l == 1
+      eat_nl()
+      local is_end = l == #lines
+      if not (is_code_block(lines[l + 1]) or is_rule(lines[l + 1]) or is_start or is_end) then
+        table.insert(ret, { line = '' })
+      end
+    elseif is_code_block(line) then
+      local lang = line:match('```%s*(%S+)') or opts.ft or 'text'
+      local block = { lang = lang, code = {} }
+      while lines[l + 1] and not is_code_block(lines[l + 1]) do
+        table.insert(block.code, lines[l + 1])
+        l = l + 1
+      end
+
+      local prev = ret[#ret]
+      if prev and not is_rule(prev.line) then
+        table.insert(ret, { line = '' })
+      end
+
+      table.insert(ret, block)
+      l = l + 1
+      eat_nl()
+    elseif is_rule(line) then
+      table.insert(ret, { line = '---' })
+      eat_nl()
+    else
+      local prev = ret[#ret]
+      if prev and prev.code then
+        table.insert(ret, { line = '' })
+      end
+      table.insert(ret, { line = line })
+    end
+    l = l + 1
+  end
+
+  return ret
+end
+
+---@type table<string, Query?>
+local ts_queries = {}
+---@param lang string
+local function get_ts_query(lang)
+  if not ts_queries[lang] then
+    ts_queries[lang] = vim.treesitter.query.get(lang, 'highlights')
+  end
+  return ts_queries[lang]
+end
+
+local md_namespace = api.nvim_create_namespace('vim_lsp_markdown')
+
+---@param bufnr integer
+---@param lang string
+---@param range Range4
+local function treesitter_highlight(bufnr, lang, range)
+  -- We can't use a cached parser here since that could interfer with the existing parser of the buffer.
+  -- TODO: ^^^ Comment from noice, what do you mean folke?
+  local LanguageTree = require('vim.treesitter.languagetree')
+  local parser = LanguageTree.new(bufnr, lang, { injections = { php = '', html = '' } })
```

Comment:
Do we exactly need to disable php/html injections for these nodes. If php and html are to be disabled, the question is: why not other parsers as well

Reply:
It seems that `noice` used to disable all injections, but that was changed in https://github.com/folke/noice.nvim/commit/0e1bf11d46054b8ab04eb62b53c5ac81b44f14df. When @folke maybe he'll tell us more about why this was needed.